single_kernel_mongo/managers/mongodb_operator.py

```diff
@@ -1047,3 +1124,19 @@ def get_statuses(self, scope: DPHScope, recompute: bool = False) -> list[StatusO
             return charm_statuses
 
         return charm_statuses
+
+    def get_password_management_statuses(self) -> list[StatusObject]:
+        """Returns the statuses related to the system-users configuration."""
+        statuses = []
+        if self.state.is_role(MongoDBRoles.SHARD) and self.config.system_users:
+            statuses.append(CharmStatuses.PASSWORD_ON_SHARD.value)
+            return statuses
+
+        if system_users_secret_id := self.config.system_users:
+            try:
+                user_passwords = self.charm.state.get_secret_from_id(system_users_secret_id)
+                validate_charm_user_password_config(user_passwords)
+            except (ModelError, SecretNotFoundError, InvalidPasswordError) as e:
+                logger.error(f"Invalid system-users secret: {e}.")
+                statuses.append(CharmStatuses.INVALID_SYSTEM_USERS.value)
```

Comment:
As I wrote above:
 * Same errors should lead to same statuses, so you need to align between `leader-elected`, `config-changed` and this piece of code.
 * I have the feeling that this code duplication highlights an issue in the implementation: I would advise you to have a look at other managers where we were facing the same kind of issues, such as the `BackupManager` or `LdapManager`: there, we build a `state` object based on the check, that is then mapped according to the function to state value: set a status + raise / set a status or map directly the state to a status.

This will bring a much more cleaner code and an easier one to understand.